File: bridge/src/bridge_store.cpp
#include <stdio.h>

#if defined(DM_PLATFORM_HTML5)
#include "bridge.h"
#include "bridge_store.h"
#include <dmsdk/dlib/crypt.h>

namespace { // private
    void cpp_bridge_store_get(dmScript::LuaCallbackInfo* onSuccess,
                                     dmScript::LuaCallbackInfo* onFailure,
                                     const int callbackType, char* dataOrError) {
        if (callbackType == 0 && dmScript::IsCallbackValid(onSuccess)) {
            lua_State* L = dmScript::GetCallbackLuaContext(onSuccess);
            if (dmScript::SetupCallback(onSuccess)) {
                uint32_t dataLen = strlen(dataOrError);
                uint32_t dstlen = dataLen * 3 / 4;
                uint8_t* dst = (uint8_t*)malloc(dstlen);
                if (!dmCrypt::Base64Decode((const uint8_t*)dataOrError, dataLen, dst, &dstlen)) {
                    free(dst);
                    dmLogError("Can't decode Base64 string.");
                }
                dmScript::PushTable(L, (const char*)dst, dstlen);
                dmScript::PCall(L, 2, 0);
                dmScript::TeardownCallback(onSuccess);
            } else {
                dmLogError("Failed to setup onSuccess");
            }
        }

        if (callbackType == 1 && dmScript::IsCallbackValid(onFailure)) {
            lua_State* L = dmScript::GetCallbackLuaContext(onFailure);
            if (dmScript::SetupCallback(onFailure)) {
                lua_pushstring(L, dataOrError);
                dmScript::PCall(L, 2, 0);
                dmScript::TeardownCallback(onFailure);
            } else {
                dmLogError("Failed to setup onOpen");
            }
        }

        if (dmScript::IsCallbackValid(onSuccess)) {
            dmScript::DestroyCallback(onSuccess);
        }
        if (dmScript::IsCallbackValid(onFailure)) {
            dmScript::DestroyCallback(onFailure);
        }
        free(dataOrError);
    }

    void cpp_bridge_store_set(dmScript::LuaCallbackInfo* onSuccess,
                              dmScript::LuaCallbackInfo* onFailure,
                              const int callbackType, char* error) {
        if (callbackType == 0 && dmScript::IsCallbackValid(onSuccess)) {
            lua_State* L = dmScript::GetCallbackLuaContext(onSuccess);
            if (dmScript::SetupCallback(onSuccess)) {
                dmScript::PCall(L, 1, 0);
                dmScript::TeardownCallback(onSuccess);
            } else {
                dmLogError("Failed to setup onSuccess");
            }
        }

        if (callbackType == 1 && dmScript::IsCallbackValid(onFailure)) {
            lua_State* L = dmScript::GetCallbackLuaContext(onFailure);
            if (dmScript::SetupCallback(onFailure)) {
                lua_pushstring(L, error);
                dmScript::PCall(L, 2, 0);
                dmScript::TeardownCallback(onFailure);
            } else {
                dmLogError("Failed to setup onOpen");
            }
        }

        if (dmScript::IsCallbackValid(onSuccess)) {
            dmScript::DestroyCallback(onSuccess);
        }
        if (dmScript::IsCallbackValid(onFailure)) {
            dmScript::DestroyCallback(onFailure);
        }
        free(error);
    }

    void cpp_bridge_store_delete(dmScript::LuaCallbackInfo* onSuccess,
                                 dmScript::LuaCallbackInfo* onFailure,
                                 const int callbackType,
                                 char* error) {
        if (callbackType == 0 && dmScript::IsCallbackValid(onSuccess)) {
            lua_State* L = dmScript::GetCallbackLuaContext(onSuccess);
            if (dmScript::SetupCallback(onSuccess)) {
                dmScript::PCall(L, 1, 0);
                dmScript::TeardownCallback(onSuccess);
            } else {
                dmLogError("Failed to setup onSuccess");
            }
        }

        if (callbackType == 1 && dmScript::IsCallbackValid(onFailure)) {
            lua_State* L = dmScript::GetCallbackLuaContext(onFailure);
            if (dmScript::SetupCallback(onFailure)) {
                lua_pushstring(L, error);
                dmScript::PCall(L, 2, 0);
                dmScript::TeardownCallback(onFailure);
            } else {
                dmLogError("Failed to setup onOpen");
            }
        }

        if (dmScript::IsCallbackValid(onSuccess)) {
            dmScript::DestroyCallback(onSuccess);
        }
        if (dmScript::IsCallbackValid(onFailure)) {
            dmScript::DestroyCallback(onFailure);
        }
        free(error);
    }
} // namespace

void bridge::store::get(const char* key, dmScript::LuaCallbackInfo* onSuccess,
                        dmScript::LuaCallbackInfo* onFailure, const char* storageType) {
    js_bridge_store_get((StoreGetHandler)cpp_bridge_store_get, key, onSuccess, onFailure, storageType);
}

void bridge::store::set(const char* key, const char* value,
                        dmScript::LuaCallbackInfo* onSuccess,
                        dmScript::LuaCallbackInfo* onFailure, const char* storageType) {
    js_bridge_store_set((StoreSetHandler)cpp_bridge_store_set, key, value, onSuccess, onFailure, storageType);
}

bool bridge::store::isSupported(const char* storageType) { return js_bridge_store_isSupported(storageType); }

bool bridge::store::isAvailable(const char* storageType) { return js_bridge_store_isAvailable(storageType); }

char* bridge::store::defaultType() { return js_bridge_store_defaultType(); }

void bridge::store::deleteData(const char* key, dmScript::LuaCallbackInfo* onSuccess,
                               dmScript::LuaCallbackInfo* onFailure, const char* storageType) {
    js_bridge_store_delete((StoreDeleteHandler)cpp_bridge_store_delete, key, onSuccess, onFailure, storageType);
}

#endif

File: bridge/lib/web/bridge_store.js
let js_bridge_store = {
    js_bridge_store_defaultType() {
        return stringToNewUTF8(bridge.storage.defaultType);
    },

    js_bridge_store_isAvailable(storageType) {
        return bridge.storage.isAvailable(UTF8ToString(storageType));
    },
    
    js_bridge_store_isSupported(storageType) {
        return bridge.storage.isSupported(UTF8ToString(storageType));
    },

    js_bridge_store_get: function (handler, key, success, failure, storageType) {
        bridge.storage.get(UTF8ToString(key), UTF8ToString(storageType))
            .then(data => {
                var luaData = "";
                if(Array.isArray(data)) {
                    luaData = data[0];
                } else {
                    luaData = data;
                }
                {{{ makeDynCall('viiii', 'handler') }}} (success, failure, 0, stringToNewUTF8(luaData));
            })
            .catch(error => {
                {{{ makeDynCall('viiii', 'handler') }}} (success, failure, 1, stringToNewUTF8("" + error));
            });
    },

    js_bridge_store_set: function (handler, key, value, success, failure, storageType) {
        bridge.storage.set(UTF8ToString(key), UTF8ToString(value), UTF8ToString(storageType))
            .then(() => {
                {{{ makeDynCall('viiii', 'handler') }}} (success, failure, 0, stringToNewUTF8(""));
            })
            .catch(error => {
                {{{ makeDynCall('viiii', 'handler') }}} (success, failure, 1, stringToNewUTF8("" + error));
            });
    },

    js_bridge_store_delete: function (handler, key, success, failure, storageType) {
        bridge.storage.delete(UTF8ToString(key), UTF8ToString(storageType))
            .then(() => {
                {{{ makeDynCall('viiii', 'handler') }}} (success, failure, 0, stringToNewUTF8(""));
            })
            .catch(error => {
                {{{ makeDynCall('viiii', 'handler') }}} (success, failure, 1, stringToNewUTF8("" + error));
            });
    }
}

mergeInto(LibraryManager.library, js_bridge_store);